app/gui/view/graph-editor/src/component/visualization/foreign/java_script/source.rs

```diff
@@ -0,0 +1,165 @@
+            files:              default(),
+        }
+    }
+
+    /// Constructor from a list of `(file_name, file_content)` pairs.
+    ///
+    /// `file_name` should be a file name or a relative path to the file.
+    pub fn from_files(files: &[(&str, &str)]) -> Self {
+        let mut sources = Self::new();
+        for (path, content) in files {
+            sources.add_file(path, content);
+        }
+        sources
+    }
+
+    /// Add a new file to the sources set, concatenating it with other files and updating source
+    /// map information.
+    pub fn add_file(&mut self, path: &str, content: &str) {
+        let source_id = self.source_map_builder.add_source(path);
+        self.source_map_builder.set_source_contents(source_id, Some(content));
+        let added_lines_count = content.lines().count() as u32;
+        for line in 0..added_lines_count {
+            // This offset is needed because of how DevTools handle source maps for
+            // `eval`uated code. Our visualizations source code is turned into a JS object by
+            // constructing a `Function` object
+            // (see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function).
+            // DevTools basically wrap provided source code with `function anonimous(<args>) {
+            // <sources> }`, so the actual line numbers in resulted code are off by 2. This was
+            // calculated empirically, and apparently, this behavior of DevTools is undocumented.
+            const OFFSET: u32 = 2;
+            const LINE_START: u32 = 0;
+
+            let dst_line = self.lines_count + line + OFFSET;
+            let source = Some(source_id);
+            self.source_map_builder.add_raw(dst_line, LINE_START, line, LINE_START, source, None);
+        }
+        self.lines_count += added_lines_count;
+        self.files.push((PathBuf::from(path), source_id));
+        self.code += content;
+    }
+
+    /// Get the final source string consisted of concatenated files and an inlined source map.
+    pub fn to_string(mut self, project: &Project) -> String {
+        self.set_file_names(project);
+        let encoded_source_map = Self::source_map_to_string(self.source_map_builder);
+        const SOURCE_MAPPING_URL: &str =
+            "\n//# sourceMappingURL=data:application/json;charset=utf-8;base64,";
+        [&self.code, SOURCE_MAPPING_URL, &encoded_source_map].concat()
+    }
+
+    /// Set file names to follow `<path_prefix>/<original_file_name>` pattern. See `path_prefix`
+    /// module documentation.
+    fn set_file_names(&mut self, project: &Project) {
+        let path_prefix = path_prefix::from_project(project);
+        for (path, source_id) in &self.files {
+            let source = path_prefix.join(path);
+            self.source_map_builder.set_source(*source_id, &source.to_string_lossy());
+        }
+    }
+
+    /// Convert source map into a base64-encoded JSON for further inlining.
```

Comment:
What does "further inlining" mean here? Is this meant to compress the content? If so, the method name could be more descriptive as the difference between `to_string` and `source_map_to_string` is a bit unclear at first glance.

Reply:
This is a helper function (hence it's private) to generate a base64-encoded source map which will later be concatenated to the actual sources (this is called inlining, as the opposite of providing source maps in separate files). I changed the name of the method to `source_map_to_base64` and reworded the doc string a bit. Is it more clear now?